### PR TITLE
Ability to set font size and color in setOSD and createOSD methods.

### PR DESCRIPTION
**Pull request by RotemDoar**
I implemented the capability to adjust font size and color within both the 'setOSD' and 'createOSD' methods. 

While working, I recognized the necessity to customize font attributes like size and color in the screen display. 

Upon reviewing the Onvif MediaBinding spec (both media and media2 versions) - 
https://www.onvif.org/ver10/media/wsdl/media.wsdl
https://www.onvif.org/ver20/media/wsdl/media.wsdl
I found that in the setOSD and createOSD function it is possible to set the font color and size.

To adding the ability, I introduced new parameters to the method options for setting font size and color in the 'setOSD' and 'createOSD' functions defined in media.js file:

```
* @param {number} [options.fontSize] The text font size.
* @param {object} [options.fontColor] The color of the text font (OSDColor), should be object with properties - X, Y, Z.
```
And in the request body as part of the TextString I added : 
```
${
	options.fontSize ?
		`<sch:FontSize>${options.fontSize}</sch:FontSize>` 
         : ''
}				
${
	options.fontColor ? 
		`
		<sch:FontColor>
		${'<sch:Color Z="' + options.fontColor.Z + '" Y="' + options.fontColor.Y + '" X="' + options.fontColor.X + '"/>'}
                </sch:FontColor>`
        : ''
}
```

Example of request options while I run the changes for one of my cameras - 
```
{
          videoSourceConfigurationToken: "VideoSourceConfigToken-01-0",
          OSDToken: "OSDDateTimeToken-0",
          position: "LowerLeft",
          timeFormat: "HH:mm:ss",
          dateFormat: "YYYY-MM-DD",
          fontSize: 1,
          fontColor: {
            X: 82,
            Y: 90,
            Z: 240,
          }
} 
```
As a result the font color configuration (X,Y,Z) changed the OSD color to red with minimum size of the font .

Moreover, I enhanced the 'create OSD' method to include the functionality of setting custom position, date, and time, aligning with the capabilities offered in the 'set OSD' method.

### CODE REVIEW

File: lib/media.js [hunk 1]
```diff
@@ -1259,21 +1265,31 @@ module.exports = function(Cam) {
...
+		`<sch:Type>DateAndTime</sch:Type>
+									<sch:DateFormat>${options.dateFormat}</sch:DateFormat>
+									<sch:TimeFormat>${options.timeFormat}</sch:TimeFormat>`
+		: `<sch:Type>Plain</sch:Type>
+									<sch:PlainText>${options.plaintext}</sch:PlainText>`}
+
+					${options.fontSize ? `<sch:FontSize>${options.fontSize}</sch:FontSize>` : ""}				
+					${options.fontColor ? `
+						<sch:FontColor>
+				        ${'<sch:Color Z="' + options.fontColor.Z + '" Y="' + options.fontColor.Y + '" X="' + options.fontColor.X + '"/>'}
```
Comment (agsh) [inline lib/media.js:1287]: Maybe add `Colorspace="http://www.onvif.org/ver10/colorspace/RGB"` here? Or add optional param to choose between YCbCr and RGB?

File: lib/media.js [hunk 2]
```diff
@@ -1245,11 +1245,17 @@ module.exports = function(Cam) {
...
-	 * @param {string} [options.position] UpperLeft, UpperRight, LowerLeft or LowerRight. Default LowerLeft   (custom mode currently not implemented)
+	 * @param {object|string} [options.position] String options: UpperLeft, UpperRight, LowerLeft or LowerRight. Default LowerLeft. Or an object with x and y position
+	 * @param {number} [options.position.x] x position of OSD, range: -1 to 1, counting from left to right
+	 * @param {number} [options.position.y] y position of OSD, range: -1 to 1, counting from up to down
+	 * @param {string} [options.plaintext] Plain text to overlay
+	 * @param {string} [options.dateFormat] Date to overlay. Must be used with timeFormat, otherwise plaintext will be used.
+	 * @param {string} [options.timeFormat] Time to overlay. Must be used with dateFormat, otherwise plaintext will be used.
+	 * @param {number} [options.fontSize] The text font size.
+	 * @param {object} [options.fontColor] The color of the text font (OSDColor), should be object with properties - X, Y, Z.
 	 * @param {Cam~GetOSDOptionsCallback} callback
```
Comment (agsh) [inline lib/media.js:1259]: You can add your example here:
```
* @example
* await cam.createOSD({
*           videoSourceConfigurationToken: "VideoSourceConfigToken-01-0",
*           OSDToken: "OSDDateTimeToken-0",
*           position: "LowerLeft",
*           timeFormat: "HH:mm:ss",
*           dateFormat: "YYYY-MM-DD",
*           fontSize: 1,
*           fontColor: {
*             X: 82,
*             Y: 90,
*             Z: 240,
*           }
* });
```

File: lib/media.js [hunk 3]
```diff
@@ -1293,6 +1310,8 @@ module.exports = function(Cam) {
    * @param {string} [options.plaintext] Plain text to overlay
    * @param {string} [options.dateFormat] Date to overlay. Must be used with timeFormat, otherwise plaintext will be used.
    * @param {string} [options.timeFormat] Time to overlay. Must be used with dateFormat, otherwise plaintext will be used.
+   * @param {number} [options.fontSize] The text font size.
+   * @param {object} [options.fontColor] The color of the text font (OSDColor), should be object with properties - X, Y, Z.
    * @param {Cam~GetOSDOptionsCallback} callback
```
Comment (agsh) [inline lib/media.js:1315]: And here link to createOSD method:
```
* @see {Cam~createOSD}
```

File: lib/media.js (hunk 1 above)
Comment (RotemDoar) [reply lib/media.js]: Added new options parameter - colorspace.
Supporting YCbCr and RGB.

File: lib/media.js (hunk 2 above)
Comment (RotemDoar) [reply lib/media.js]: Added example to create osd.

File: lib/media.js (hunk 3 above)
Comment (RotemDoar) [reply lib/media.js:1339]: Added the link to create osd method.

**Comment by RotemDoar** [comment]
@agsh 

I have added an optional parameter - 'colorspace', allowing support for both YCbCr and RGB. The default setting is RGB.
(see line 1259 in createOSD and line 1334 in setOSD).

`* @param {string} [options.colorspace] Colorspace - RGB or YCbCr. Default RGB.
`

and fix the request for the support (with extra validation to the parameters X,Y,Z) - 
(line 1304 and 1369)

```
${ options.fontColor && options.fontColor.X && options.fontColor.Y && options.fontColor.Z ? 
		`
		<sch:FontColor>
		${ '<sch:Color Z="' + options.fontColor.Z + '" Y="' + options.fontColor.Y + '" X="' + options.fontColor.X + '" Colorspace="' + `${ options.colorspace === "YCbCr" ? "http://www.onvif.org/ver10/colorspace/YCbCr" : "http://www.onvif.org/ver10/colorspace/RGB" }` + '"/>' }
		</sch:FontColor>` : "" }
```

I added extra documentation to the font color parameter.

```
* @param {object} [options.fontColor] The color of the text font (OSDColor), should be object with properties - X, Y, Z.
* @param {float} [options.fontColor.X] For RGB means R value, For YCbCr means Y value.
* @param {float} [options.fontColor.Y] For RGB means G value, For YCbCr means Cb value.
* @param {float} [options.fontColor.Z] For RGB means B value, For YCbCr means Cr value.
```

I added example to the parameters for createOSD method.

```
* await cam.createOSD({
   *           position: "LowerLeft",
   *           timeFormat: "HH:mm:ss",
   *           dateFormat: "YYYY-MM-DD",
   *           fontSize: 1,
   *           colorspace: "RGB",
   *           fontColor: {
   *             X: 1,
   *             Y: 0.7,
   *             Z: 0.9,
   *           }
   * });
```

I added the link to create osd method in the setOSD method.

`   * @see {Cam~createOSD} 
`

**Review by agsh — APPROVED** [review]

**Comment by agsh** [comment]
@RotemDoar Great, thanks! Nice and detailed comments :smile_cat:

**Pull request merged**